### PR TITLE
[MCC-483] TxManager trait

### PR DESCRIPTION
**Pull request by mfaulk**
Soundtrack of this PR: [All About That Trait](https://www.youtube.com/watch?v=7PCkvCPvDXk)

### Motivation

Creates a trait for TxManager so that it can be mocked with mockall.

### In this PR
* Splits TxManager into a few smaller files. I find that easier to navigate.
* Creates a trait and mock. Changes a method to avoid a higher-ranked trait bound, which mockall doesn't support.

### Future Work
* Update ByzantineLedger, ConsensusService, ClientApiService, PeerApiService, etc. to depend on the trait instead of the struct, so that they can be tested using the mock.
* Will probably rename TxManager and TxManagerTrait to TxManagerImpl and TxManager.

### CODE REVIEW

File: consensus/service/src/byzantine_ledger/worker.rs [hunk 1]
```diff
@@ -610,33 +610,16 @@ impl<
...
-            log::debug!(
-                self.logger,
-                "attempting to resolve missing tx hashes ({}, left with {}): {:?}",
-                missing_hashes.len(),
-                all_missing_hashes.len(),
-                missing_hashes,
-            );
-
-            match conn.fetch_txs(&missing_hashes, Fibonacci::from_millis(100).take(10)) {
+        for chunk in missing_hashes[..].chunks(100) {
```
Comment (mfaulk) [inline consensus/service/src/byzantine_ledger/worker.rs:613]: `chunks` makes this easier.

File: consensus/service/src/tx_manager/mod.rs [hunk 2]
```diff
@@ -223,25 +158,18 @@ impl<E: ConsensusEnclave, UI: UntrustedInterfaces> TxManager<E, UI> {
         expired.keys().cloned().collect()
     }
 
-    /// Returns elements of `tx_hashes` that are not inside the cache.
-    pub fn missing_hashes<T>(&self, tx_hashes: &T) -> Vec<TxHash>
-    where
-        for<'a> &'a T: IntoIterator<Item = &'a TxHash>,
```
Comment (mfaulk) [inline consensus/service/src/tx_manager/mod.rs:229]: mockall doesn't like this higher-ranked trait bound (it lets you pass a reference to any iterable collection of TxHashes as a function argument). I'm guessing that the simpler `contains` will work OK, but I will also check if the extra locking causes performance problems.

File: transaction/core/src/onetime_keys.rs [hunk 3]
```diff
@@ -121,7 +121,7 @@ pub fn view_key_matches_output(
 /// * `tx_public_key` - The transaction public key `R`.
 /// * `view_private_key` - A private view key `a`.
 /// * `subaddress_spend_private_key` - A private spend key `d = Hs(a || i) + b`.
-///
+///69
```
Comment (mfaulk) [inline transaction/core/src/onetime_keys.rs:124]: I blame my cat. Will fix.

File: consensus/service/src/tx_manager/mod.rs [hunk 4]
```diff
@@ -7,72 +7,29 @@
 //! may operate on.
 
 use crate::counters;
-use failure::Fail;
+
```
Comment (eranrund) [inline consensus/service/src/tx_manager/mod.rs:10]: We generally don't split imports so this empty line isn't necessary.

File: consensus/service/src/tx_manager/mod.rs (hunk 2 above)
Comment (eranrund) [reply consensus/service/src/tx_manager/mod.rs:229]: Yup, would be interesting to see what effect this has. I think this is going to move us from one lock per SCP statement, so N locks per SCP statement, where N is the number of tx hashes the statement refers to.

File: transaction/core/src/onetime_keys.rs (hunk 3 above)
Comment (eranrund) [reply transaction/core/src/onetime_keys.rs:124]: 🐈 🐈 🐈 🐈

File: consensus/service/src/tx_manager/error.rs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+use failure::Fail;
```
Comment (rjwalters) [inline consensus/service/src/tx_manager/error.rs:3]: deprecated?

File: consensus/service/src/tx_manager/mod.rs (hunk 2 above)
Comment (rjwalters) [reply consensus/service/src/tx_manager/mod.rs:229]: https://stackoverflow.com/questions/3652056/how-efficient-is-locking-an-unlocked-mutex-what-is-the-cost-of-a-mutex

should we place bets? :)

File: consensus/service/src/tx_manager/error.rs (hunk 5 above)
Comment (mfaulk) [reply consensus/service/src/tx_manager/error.rs:3]: I'll pick that up on the next PR.

**Comment by mfaulk** [comment]
Performance seems unchanged:
Screenshot from 2020-08-24 20-47-23

**Pull request merged**